[STORM-1175] State store for windowing operations

arunmahadevan opened the pull request:
Added IStatefulBolt abstraction that can be implemented by bolts which requires
its state to be checkpointed periodically. State implementations based on
key-value mapping store are added. There is a default in-memory based implementation
and optional redis based implementation that provides state persistence. An internal
CheckpointSpout periodically emits checkpoint tuples which flows through the
topology DAG to take a consistent snapshot of the state across all components.

There is still pending work to capture the evaluated/expired state of the tuples
in the Window and use it to prune duplicate window evaluations during restart. This can
be built on top of the stateful bolt abstractions and will be done as part of STORM-1176

hmcl commented on storm-core/test/jvm/backtype/storm/state/InMemoryKeyValueStateTest.java:
```diff
@@ -0,0 +1,73 @@
+        keyValueState.put("b", "2");
+        keyValueState.prepareCommit(1);
+        keyValueState.put("c", "3");
+        assertEquals("1", keyValueState.get("a"));
```
Please put all of this duplicated code in a helper method. All of this copy and paste should be one line helper method call

harshach commented on external/storm-redis/src/main/java/org/apache/storm/redis/state/RedisKeyValueState.java:
```diff
@@ -0,0 +1,302 @@
+ */
+public class RedisKeyValueState<K, V> implements KeyValueState<K, V> {
+    private static final Logger LOG = LoggerFactory.getLogger(RedisKeyValueState.class);
+    private static final String COMMIT_TXID_KEY = "$commit";
```
any reason for using $ notation here.

arunmahadevan replied:
Since its internal, just followed a convention to differentiate it (similar to checkpoint stream). But here there is no namespace collision so it could be "commit" as well.

harshach commented on storm-core/src/jvm/backtype/storm/Config.java:
```diff
@@ -1491,6 +1491,29 @@
+     * This can be overridden at the component level.
+     */
+    @isString
+    public static final String TOPOLOGY_STATE_PROVIDER_CONFIG = "topology.state.provider.config";
```
whats is this config suppose to do ? will it accept it file that contains config?

arunmahadevan replied:
For now we expect the actual config string (for redis, its a json string with the host, port etc). it could be a file name as well, if the state provider implementation desires so.

harshach commented on storm-core/src/jvm/backtype/storm/spout/CheckPointState.java:
```diff
@@ -0,0 +1,78 @@
+        /**
+         * The checkpoint spout has started committing the transaction.
+         */
+        COMMITTING,
```
should we call this as COMMIT_IN_PROGRESS

arunmahadevan replied:
Named it COMMITTING and PREPARING since its short. We could probably have it as COMMIT_IN_PROGRESS and PREPARE_IN_PROGRESS. Prefer to keep it short and can add more comments.

harshach commented on storm-core/src/jvm/backtype/storm/spout/CheckPointState.java:
```diff
@@ -0,0 +1,78 @@
+        /**
+         * The checkpoint spout has started preparing the transaction for commit.
+         */
+        PREPARING
```
PREPARE probably better

arunmahadevan replied:
It could be PREPARE_IN_PROGRESS or PREPARING, I think PREPARE conveys a slightly different meaning.

Parth-Brahmbhatt commented on storm-core/src/jvm/backtype/storm/spout/CheckPointState.java:
```diff
@@ -0,0 +1,78 @@
+ * {@link CheckpointSpout}
+ */
+public class CheckPointState {
+    public long txid;
```
shouldn't these be private fields with getters?

arunmahadevan replied:
will change it

harshach commented on storm-core/src/jvm/backtype/storm/spout/CheckpointSpout.java:
```diff
@@ -0,0 +1,280 @@
+    public static final String CHECKPOINT_ACTION_INITSTATE = "initstate";
+
+    private static final String TX_STATE_KEY = "__state";
+    private static final int DEFAULT_CHECKPOINT_INTERVAL = 1000; // every sec
```
can we have this as part of defaults.yaml

arunmahadevan replied:
Its used only if Config.TOPOLOGY_STATE_CHECKPOINT_INTERVAL is not set. Will add comments to make it clear. Will also set the default value for that config in defaults.yaml

harshach replied:
@arunmahadevan well if you add that to defaults.yaml it will be set and thats how the defaults has been used so far in storm. Incase if we need to change default in the future its .yaml file change than the code change.

Parth-Brahmbhatt commented on storm-core/src/jvm/backtype/storm/spout/CheckpointSpout.java:
```diff
@@ -0,0 +1,280 @@
+ * <p/>
+ * <pre>
+ *                  ROLLBACK(tx2)
+ *               <-------------                  PREPARE(tx2)                     COMMIT(tx2)
```
why Rollback can only happen from Prepare and not from Committing? in that case why do we have Committing state?

arunmahadevan replied:
Rollback can happen from 'preparing' state since all bolts might not have prepared the data to commit. Once we get ack for prepare it indicates all bolts have prepared the data to commit, the state changes from 'preparing' to 'committing' and a 'commit' message is sent. We cant rollback from 'committing' state since some bolts might have already committed the data. In case of failures, the commit message is sent again to make sure all bolts commit. If some bolts already committed the prepared data, it ignores the message.

Will add more comments to make it clear.

Parth-Brahmbhatt commented on storm-core/src/jvm/backtype/storm/spout/CheckpointSpout.java:
```diff
@@ -0,0 +1,280 @@
+        if (shouldRecover()) {
+            LOG.debug("In recovery");
+            handleRecovery();
+            startProgress();
```
given you do this in both if and else as the last step, probably just move it outside.

arunmahadevan replied:
Actually it can't be moved out since its in the `elseif`. If nextTuple is invoked while its in the middle of recovering/checkpointing, it does not do anything. I will add a final else and put logs/comment.

Parth-Brahmbhatt commented on storm-core/src/jvm/backtype/storm/spout/CheckpointSpout.java:
```diff
@@ -0,0 +1,280 @@
+    @Override
+    public Map<String, Object> getComponentConfiguration() {
+        Config conf = new Config();
+        conf.put(Config.TOPOLOGY_SLEEP_SPOUT_WAIT_STRATEGY_TIME_MS, 100);
```
anyway to override this?

arunmahadevan replied:
It was added to sleep for slightly longer time rather than spinning in the nextTuple. Instead of hard coding, will change this to 10% of the checkpoint interval time.

Parth-Brahmbhatt commented on storm-core/src/jvm/backtype/storm/spout/CheckpointSpout.java:
```diff
@@ -0,0 +1,280 @@
+        startProgress();
+    }
+
+    private void handleRecoveryAck() {
```
the state transition code seems to be repeating in handleCheckpointAck, doCheckpoint and handleRecoveryAck. Do you think we can capture the nextState and prevState for each state in the CheckPointState class it self so in all these methods we can just call netxt/prevState followed by saveTxState(txState).

arunmahadevan replied:
Yes it will make it cleaner. I will try to refactor it and move it to nextAction and nextState in checkpoint state.

Parth-Brahmbhatt commented:
@arunmahadevan Thanks for the patch. I think the missing piece is the how-to-use guide. you can file a follow up jira for that. Overall looks good to me and once the comments are addressed , I am +1.

arunmahadevan commented:
@Parth-Brahmbhatt @harshach refactored the state handling code to make it cleaner and addressed other review comments. Also added a followup task for documentation - https://issues.apache.org/jira/browse/STORM-1416

harshach commented:
+1 as well. minor nit pick on the default value.

arunmahadevan commented:
@harshach while addressing review comments I had added the config in defaults.yaml. I further refactored it to remove the constant and made sure that the value is not too less.

arunmahadevan commented:
@harshach @Parth-Brahmbhatt whilde doing more tests found out a few issues around acking and failure handling and have addressed in https://github.com/apache/storm/commit/c075ffacdf1680ea3792c4e7979228e3a1d65f0c can you take a look?

Parth-Brahmbhatt commented:
@arunmahadevan I am still +1 but the up merge is failing for storm-core/src/jvm/org/apache/storm/topology/TopologyBuilder.java can you please upmerge the last time and I can merge this into master.

arunmahadevan commented:
@Parth-Brahmbhatt Fixed the merge conflicts. 

Also had to refactor most of the code since the package name changed from backtype.storm -> org.apache.storm in master. Ran the sample topologies after the package name refactoring and things look good.

ptgoetz commented:
+1

arunmahadevan commented:
@kosii Thanks for reporting. Yes the keys in `pendingPrepare` should have priority. Typically this case does not happen since when a prepare fails, the txn is rolled-back and retried (i.e a fresh prepare). 

Do you want to post a JIRA/Patch?

kosii commented:
@arunmahadevan yes, I'm on it
